app/src/main/java/fr/neamar/kiss/result/ShortcutsResult.java

```diff
@@ -104,19 +117,50 @@ public Drawable getDrawable(Context context) {
 
     @Override
     protected void doLaunch(Context context, View v) {
-
-        try {
-            Intent intent = Intent.parseUri(shortcutPojo.intentUri, 0);
-            if (android.os.Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR2) {
-                intent.setSourceBounds(v.getClipBounds());
+        if (shortcutPojo.isOreoShortcut()) {
+            // Oreo shortcuts
+            doOreoLaunch(context, v);
+        } else {
+            // Pre-oreo shortcuts
+            try {
+                Intent intent = Intent.parseUri(shortcutPojo.intentUri, 0);
+                if (android.os.Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR2) {
+                    intent.setSourceBounds(v.getClipBounds());
+                }
+
+                context.startActivity(intent);
+            } catch (Exception e) {
+                // Application was just removed?
+                Toast.makeText(context, R.string.application_not_found, Toast.LENGTH_LONG).show();
             }
+        }
+    }
 
-            context.startActivity(intent);
-        } catch (Exception e) {
-            // Application was just removed?
-            Toast.makeText(context, R.string.application_not_found, Toast.LENGTH_LONG).show();
+    @TargetApi(Build.VERSION_CODES.O)
+    private void doOreoLaunch(Context context, View v) {
+        final LauncherApps launcherApps = (LauncherApps) context.getSystemService(Context.LAUNCHER_APPS_SERVICE);
+        assert launcherApps != null;
+
+        // Only the default launcher is allowed to start shortcuts
+        if (!launcherApps.hasShortcutHostPermission()) {
+            return;
         }
 
+        LauncherApps.ShortcutQuery query = new LauncherApps.ShortcutQuery();
+        query.setPackage(shortcutPojo.packageName);
+        query.setShortcutIds(Collections.singletonList(shortcutPojo.getOreoId()));
+        query.setQueryFlags(LauncherApps.ShortcutQuery.FLAG_MATCH_PINNED);
+
+        List<UserHandle> userHandles = launcherApps.getProfiles();
+
+        // Find the correct UserHandle, and launch the shortcut.
+        for (UserHandle userHandle : userHandles) {
+            List<ShortcutInfo> shortcuts = launcherApps.getShortcuts(query, userHandle);
+            if (shortcuts != null && shortcuts.size() > 0 && shortcuts.get(0).isEnabled()) {
+                launcherApps.startShortcut(shortcuts.get(0), v.getClipBounds(), null);
+                return;
+            }
+        }
```

Comment:
what if we didn't find the shortcut?

might happen if the app was removed maybe?